_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,321 @@
+serial nature of Bayesian sampling algorithms. Bayesian sampling treats models
+as black boxes, and tries to deduce the set of inputs that maximize the black
+box's quality. This is difficult in a distributed system because if there were
+an infinite number of workers, the Bayesian sampling algorithm wouldn't have
+time to deduce anything before initializing models. Regardless, [BOHB] has an
+intelligent way of scaling to distributed systems given Hyperband's parallel
+architecture.[^scaling]
+
+This also looks useful towards the end of the hyperparameter optimization
+according to Figure 1 from the [BOHB paper]:
+
+<img width="50%" src="/images/2020-model-selection/bohb-fig-1.png" />
+
+[^scaling]:As mentioned in Section 4.2 and illustrated in Figure 2 of [the BOHB paper][BOHB paper].
+
+There are additional useful features of Dask-ML's hyperparameter optimization,
+especially in the context of natural language processing (NLP).  [Avoiding
+repeated work] in Scikit-learn Pipelines is a huge feature when performing NLP
+because the initial stages of analysis take an excessive amount of computation
+and memory.[^openai] Currently, Dask-ML's implementation that avoid repeated
+work is completely separate from Dask-ML's Hyperband implementation --
+naturally it's future work to combine the two.
+
+The Dask-ML model selection effectively avoids unnecessary work, both with
+unpromising hyperparameters and with unnecessary repeated work. Of course, as a
+machine learning graduate student I see opportunities for future work, some of
+which are mentioned in [the conclusion]. Luckily, Dask-ML's model selection
+framework is pretty flexible so these modifications are straightforward.
+
+[the conclusion]:#conclusion
+[hyperband-paper]:https://arxiv.org/pdf/1603.06560.pdf
+[470]:https://scholar.google.com/scholar?cites=10473284631669296057&as_sdt=5,39&sciodt=0,39&hl=en
+
+[^automl]:Their implementation of Hyperband in [HpBandSter] is included in [Auto-PyTorch] and [BOAH].
+
+[^openai]:Computing [n-grams] is difficult. For OpenAI, NLP preprocessing took 8 GPU-months! ([source][gpt])
+
+[n-grams]:https://en.wikipedia.org/wiki/N-gram
+[gpt]:https://openai.com/blog/language-unsupervised/#drawbacks
+[Avoiding repeated work]:https://ml.dask.org/hyper-parameter-search.html#avoid-repeated-work
+[BOHB]:https://automl.github.io/HpBandSter/build/html/optimizers/bohb.html
+[hscv]:https://ml.dask.org/modules/generated/dask_ml.model_selection.HyperbandSearchCV.html#dask_ml.model_selection.HyperbandSearchCV
+[BOAH]:https://github.com/automl/BOAH
+[Auto-PyTorch]:https://www.automl.org/wp-content/uploads/2018/09/chapter7-autonet.pdf
+[HpBandSter]:https://github.com/automl/HpBandSter
+[Optuna]:https://medium.com/optuna/optuna-supports-hyperband-93b0cae1a137
+[keras-tune]:https://keras-team.github.io/keras-tuner/documentation/tuners/#hyperband-class
+
+> 3\. **Framework support**: tune-sklearn is used primarily for tuning
```

Comment:
I think each of these should get a header to break up the piece a bit.

Reply:
Good idea!